### PR TITLE
Introduce a new AutocompleteInput based on MUI Autocomplete

### PR DESCRIPTION
**Pull request by djhi**
- [x] Avoid breaking change for `optionText` and leverage the `inputText` prop we had in previous version

### CODE REVIEW

File: packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx [hunk 1]
```diff
@@ -23,23 +23,8 @@ describe('<AutocompleteInput />', () => {
         resource: 'users',
     };
 
-    it('should use a Downshift', () => {
```
Comment (djhi) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx:26]: Removed because we don't want to test mui autocomplete

File: packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx [hunk 2]
```diff
@@ -645,37 +632,6 @@ describe('<AutocompleteInput />', () => {
         await waitFor(() => expect(queryAllByRole('option').length).toEqual(1));
     });
 
-    it('passes options.suggestionsContainerProps to the suggestions container', () => {
```
Comment (djhi) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx:648]: mui autocomplete provide its own ways to customize the underlying components

File: packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx [hunk 3]
```diff
@@ -694,60 +650,6 @@ describe('<AutocompleteInput />', () => {
         });
     });
 
-    it('should not render a LinearProgress if loading is true and a second has not passed yet', () => {
```
Comment (djhi) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx:697]: Mui autocomplete has its own loading mechanism. Maybe we should check we don't its loading prop unless a second has passed

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 4]
```diff
@@ -1,193 +1,63 @@
...
     UseChoicesOptions,
-    mergeRefs,
     useSuggestions,
+    useTimeout,
     useTranslate,
     warning,
 } from 'ra-core';
-
-import { InputHelperText } from './InputHelperText';
-import { AutocompleteSuggestionList } from './AutocompleteSuggestionList';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:33]: I assume that these are still required for AutocompleteArrayInput, and that's why you didn't delete them in this PR, right?

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 5]
```diff
@@ -1,193 +1,63 @@
...
-import { AutocompleteSuggestionItem } from './AutocompleteSuggestionItem';
-import { AutocompleteInputLoader } from './AutocompleteInputLoader';
 import {
     SupportCreateSuggestionOptions,
     useSupportCreateSuggestion,
 } from './useSupportCreateSuggestion';
+import { InputHelperText } from './InputHelperText';
 
-/**
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:41]: Please keep a JSDoc section

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 6]
```diff
@@ -1,193 +1,63 @@
...
         id: idOverride,
         input: inputOverride,
-        inputText,
         isRequired: isRequiredOverride,
         label,
-        limitChoicesToValue,
         loaded,
         loading,
-        margin = 'dense',
+        limitChoicesToValue,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:39]: This one needs documentation

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 7]
```diff
@@ -1,193 +1,63 @@
...
         helperText,
         id: idOverride,
         input: inputOverride,
-        inputText,
         isRequired: isRequiredOverride,
         label,
-        limitChoicesToValue,
         loaded,
         loading,
-        margin = 'dense',
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:128]: You should still support it, or it's a breaking change

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 8]
```diff
@@ -1,193 +1,63 @@
...
         resource,
         setFilter,
-        shouldRenderSuggestions: shouldRenderSuggestionsOverride,
+        shouldRenderSuggestions,
         source,
         suggestionLimit,
-        translateChoice = true,
+        translateChoice,
         validate,
-        variant = 'filled',
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:157]: you should still support it, or it's a breaking change

**Comment by fzaninotto** [comment]
I see a UI bug in the Storybook: in the "inside Reference Input" srtory, when typing text, the choices are briefly emptied before being filled with matching choices. I think you should use the keepPreviousData option in useGetList (once it's handled by react-query, of course)

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 9]
```diff
@@ -1,193 +1,63 @@
...
         ...rest
     } = props;
 
-    if (isValidElement(optionText) && !inputText) {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:161]: Why did you remove these warnings?

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx [hunk 10]
```diff
@@ -208,48 +78,34 @@ export const AutocompleteInput = (props: AutocompleteInputProps) => {
...
-        [input.value, getSuggestionFromValue]
+        () =>
+            choices?.find(choice => get(choice, optionValue) === input.value) ||
+            null,
+        [choices, input.value, optionValue]
     );
 
-    const doesQueryMatchSuggestion = useMemo(() => {
+    useEffect(() => {
         if (isValidElement(optionText)) {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:89]: why put that in an effect?

File: packages/ra-ui-materialui/src/input/SelectArrayInput.tsx [hunk 11]
```diff
@@ -135,9 +135,9 @@ export const SelectArrayInput = (props: SelectArrayInputProps) => {
     });
 
     const handleChange = useCallback(
-        (event, newItem) => {
-            if (newItem) {
-                input.onChange([...input.value, getChoiceValue(newItem)]);
+        event => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SelectArrayInput.tsx:138]: Not sure about what you did here, can you elaborate?

File: packages/ra-ui-materialui/src/input/SelectInput.tsx [hunk 12]
```diff
@@ -178,14 +178,14 @@ export const SelectInput = (props: SelectInputProps) => {
...
     const handleChange = useCallback(
-        async (event: React.ChangeEvent<HTMLSelectElement>, newItem) => {
-            if (newItem) {
-                const value = getChoiceValue(newItem);
-                input.onChange(value);
-                return;
+        async (event: any) => {
+            let choice = event;
+            // eslint-disable-next-line eqeqeq
+            if (event?.target?.value != undefined) {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/SelectInput.tsx:184]: what that code does is cryptic to me. Would you mind adding a comment explaining it?

File: packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx (hunk 3 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.spec.tsx:697]: I see that you did it, am I wrong?

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Great! 💯

File: cypress/integration/edit.js [hunk 13]
```diff
@@ -78,6 +78,12 @@ describe('Edit Page', () => {
             cy.get(EditPostPage.elements.input('backlinks[0].url')).blur();
 
             cy.contains('Required');
+            // FIXME: We navigate away from the page and confirm the unsaved changes
```
[the remark this answers is not in the document]
Comment (djhi) [reply cypress/integration/edit.js:81]: I'm not sure yet

File: examples/simple/src/comments/CommentEdit.tsx [hunk 14]
```diff
@@ -152,15 +152,19 @@ const CommentEdit = props => {
                                         suggestion
                                     ) => {
                                         const title = `${suggestion.title} - ${suggestion.id}`;
-
                                         return title.includes(filterValue);
                                     }}
-                                    optionText={<OptionRenderer />}
+                                    optionText={record =>
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/simple/src/comments/CommentEdit.tsx:157]: This is the unnecessary breaking change I was talking about. Will be fixed soon

File: examples/simple/src/comments/CommentEdit.tsx [hunk 15]
```diff
@@ -152,15 +152,19 @@ const CommentEdit = props => {
...
+                                        `${record?.title} - ${record?.id}`
+                                    }
                                     inputText={inputText}
                                     options={{
-                                        InputProps: {
-                                            fullWidth: true,
-                                        },
+                                        renderOption: (props, record) => (
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/simple/src/comments/CommentEdit.tsx:162]: This is the unnecessary breaking change I was talking about. Will be fixed soon

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx (hunk 4 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:33]: Exactly

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:39]: I actually migrated it without thinking but I fail to see its usefulness

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:161]: This is the unnecessary breaking change I was talking about. Will be fixed soon

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx (hunk 10 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:89]: Because we rely on props and we want the warning to be displayed if needed when they change?

File: examples/simple/src/comments/CommentEdit.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (djhi) [reply examples/simple/src/comments/CommentEdit.tsx:157]: Fixed

File: examples/simple/src/comments/CommentEdit.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (djhi) [reply examples/simple/src/comments/CommentEdit.tsx:162]: Fixed

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx (hunk 5 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:41]: Added

File: packages/ra-ui-materialui/src/input/AutocompleteInput.tsx (hunk 6 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/AutocompleteInput.tsx:112]: I'll revisit once the AutocompleteArrayInput is migrated to use Autocomplete

**Comment by WiXSL** [comment]
Editing a Post In simple example:

**Review by WiXSL — COMMENTED** [review]
Editing a Post In simple example:

File: UPGRADE.md [hunk 16]
```diff
@@ -755,6 +755,14 @@ test('MyComponent', () => {
...
+
+We migrated the `AutocompleteInput` so that it leverage Material UI [`<Autocomplete>`](https://mui.com/components/autocomplete/). If you relied on [Downshift](https://www.downshift-js.com/) options, you'll have to update your component.
+
+Besides, some props supported by the previous implementation aren't anymore:
+- `clearAlwaysVisible`: the clear button is now always either while hovering the input or when it has focus. You can hide it using the `<Autocomplete>` `disableClearable` prop though.
+- `resettable`: Removed for the same reason as `clearAlwaysVisible`
+
 ## `useAuthenticated` Signature has Changed
 
 `useAuthenticated` uses to accept only the parameters passed to the `authProvider.checkAuth` function. It now accept an option object with two properties:
```
Comment (WiXSL) [inline UPGRADE.md:768]:
```suggestion
`useAuthenticated` uses to accept only the parameters passed to the `authProvider.checkAuth` function. It now accepts an option object with two properties:
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by djhi** [comment]
> Editing a Post In simple example:
> 
> 
> 
> I just set a value, and then re-open the AutocompleteInput

Good catch!

File: UPGRADE.md [hunk 17]
```diff
@@ -755,9 +755,17 @@ test('MyComponent', () => {
 });
 ```
 
+## AutocompleteInput Now Uses Material UI Autocomplete
+
+We migrated the `AutocompleteInput` so that it leverage Material UI [`<Autocomplete>`](https://mui.com/components/autocomplete/). If you relied on [Downshift](https://www.downshift-js.com/) options, you'll have to update your component.
+
+Besides, some props supported by the previous implementation aren't anymore:
+- `clearAlwaysVisible`: the clear button is now always either while hovering the input or when it has focus. You can hide it using the `<Autocomplete>` `disableClearable` prop though.
```
Comment (WiXSL) [inline UPGRADE.md:763]:
```suggestion
- `clearAlwaysVisible`: the clear button is now always visible, either while hovering the input or when it has focus. You can hide it using the `<Autocomplete>` `disableClearable` prop though.
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Inputs.md [hunk 18]
```diff
@@ -662,17 +667,6 @@ import { AutocompleteInput, ReferenceInput } from 'react-admin';
 </ReferenceInput>
 ```
 
-Lastly, would you need to override the props of the suggestion's container (a `Popper` element), you can specify them using the `options.suggestionsContainerProps`. For example:
```
Comment (WiXSL) [inline docs/Inputs.md:665]: Since `suggestionsContainerProps` was introduced by react-admin, shouldn't we clarify that now, `@mui` way of overriding the `Popper` container should be used?

Comment (djhi) [reply docs/Inputs.md:665]: That's in the MUI documentation, I don't think we should include it

**Comment by WiXSL** [comment]
The [vercel deployment](https://react-admin-6xv4rxrxy-marmelab.vercel.app/#/posts) doesn't have the authors ArrayInput in Post form and lacks the User menu

**Comment by djhi** [comment]
> The [vercel deployment](https://react-admin-6xv4rxrxy-marmelab.vercel.app/#/posts) doesn't have the authors ArrayInput in Post form and lacks the User menu

You have to log in with admin/password

**Comment by WiXSL** [comment]
> I was trying it locally and it caught my attention, you are right

I was trying it locally and it caught my attention, you are right

**Review by WiXSL — APPROVED** [review]

File: UPGRADE.md [hunk 19]
```diff
@@ -755,9 +755,17 @@ test('MyComponent', () => {
 });
 ```
 
+## AutocompleteInput Now Uses Material UI Autocomplete
+
+We migrated the `AutocompleteInput` so that it leverage Material UI [`<Autocomplete>`](https://mui.com/components/autocomplete/). If you relied on [Downshift](https://www.downshift-js.com/) options, you'll have to update your component.
```
Comment (fzaninotto) [inline UPGRADE.md:760]:
```suggestion
We migrated the `AutocompleteInput` so that it leverages Material UI [`<Autocomplete>`](https://mui.com/components/autocomplete/). If you relied on [Downshift](https://www.downshift-js.com/) options, you'll have to update your component.
```

**Review by fzaninotto — APPROVED** [review]
